src/Extension/Entity/ExtensionCollection.php

```
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OAT\DependencyResolver\Extension\Entity;
+
+class ExtensionCollection extends \ArrayObject implements \JsonSerializable
+{
+    /** @var Extension[] */
+    private $extensions = [];
+
+    public function add(Extension $extension): self
+    {
+        $this->offsetSet($extension->getExtensionName(), $extension);
+
+        return $this;
+    }
+
+    public function offsetSet($index, $newval)
+    {
+        if (!$newval instanceof Extension) {
+            throw new \TypeError('Extension provided is not an instance of ' . Extension::class . '.');
+        }
+
+        $this->extensions[$index] = $newval;
+    }
+
+    public function offsetGet($index): ?Extension
+    {
+        if (!$this->offsetExists($index)) {
+            return null;
```

I would throw an exception here instead.

---

Refactored.